### DISCUSSION TITLE
Capillary hysteresis

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#23543] [Q&A Modules: General] [answered]
Hi,
Hope you are doing well.

I came across some questions, and I appreciate if someone could help me with them. Sorry if they are naive :)

I wnated to simulate an imbibtion-drainage (wetting-drying) by PorousFlow module. It supports hysteresis but the tests available are only one dimensional. So, I was wondering if this feature is supported in higher dimensions, and will it converge or not. My other question is that in the examples, hysteresis is only used with dirac kernels. Is this the only way to use hysteresis in PorousFlow? 

Thanks,
Sina

**Comment by GiudGiud** [comment]
@WilkAndy

**Comment by WilkAndy** [comment]
Yes, hysteresis can be used in models with any dimensionality.   The only reason why the examples are all 1D is because, being part of the PorousFlow test suite, they need to run very quickly.

Regarding convergence: it's quite likely that hysteresis won't impact convergence much, but whether that's true in your case depends upon your model.

I suggest that you create your model without hysteresis, get everything running nicely, and then finally include hysteresis, using the procedures described here: https://mooseframework.inl.gov/modules/porous_flow/hysteresis.html

A

**Reply by WilkAndy** [reply]
Oh, and DiracKernels are unrelated to hysteresis: the examples you looked at all happen to have DiracKernels, but they do not impact hysteresis.

**Reply by sinaomrani96** [reply]
Hi,
Hope you are doing well.

Thank you for your answer. I created the model without hysteresis based on a PorousFlow example as you suggested and it works but when I try to use PC with hysteresis I get segmentation fault error. I run the model in debugg mode and I got
 'Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 2 size: 2)
at /home/sinaomrani/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276'
I do not know why I face this error. Can you please give me your insight? Also here is the input file.
https://github.com/sinaomrani96/PorousFlow-hysteresis/blob/main/test

Thanks
Best regards,
Sina

**Reply by GiudGiud** [reply]
I took a quick look and the interaction between the PorousFlowSink and the PorousFlowHysteresis seems to have issues. 

The out-of-bounds access happens on the boundary, when computing the material property for the 3rd quadrature point when the container has only been sized for 2. I m not sure why it would not be sized well yet. @WilkAndy what are your thoughts

**Reply by WilkAndy** [reply]
I feel this is likely to be a bug!   Apologies, and kind-of-thanks (only kind-of because this means more work for me!).  I'm currently not at work, and won't be there for a few days, so can't fix it right now.  I'm not super sure why the boundary of a 2D mesh would have 3 quadpoints, but maybe something is getting evaluated on the interior element.

**Reply by sinaomrani96** [reply]
Dear @WilkAndy, 
Hope you are doing well.

I was wondering if you fixed this bug or you have not find the time yet. 

Thanks. 
Best regards,
Sina

**Reply by WilkAndy** [reply]
Sina, i'm sorry but i haven't fixed it yet.   I've been largely decoupled from MOOSE for the last month, and will not have time to work on it till after Easter.

**Reply by WilkAndy** [reply, marked as answer]
This has been fixed in https://github.com/idaholab/moose/pull/24104 , @sinaomrani96 .   Thank you for finding this bug, and I'm sorry it has impacted your work, and that i took so long to fix it.

**Comment by sinaomrani96** [comment]
Dear @WilkAndy,
Hope you are doing well.

Great. I just checked it, and it works fine. I appreciate your time. Thanks.
